Implemented autocomplete endpoint and added documentation

romain894 opened the pull request:
The library now fully supports the autocomplete endpoint. I added a section in the README with the documentation on how to use it.

I also added a section in the README to explain how to set up the config to enable the retry.

J535D165 commented on README.md:
````diff
@@ -307,6 +319,32 @@ for page in pager:
+
+Autocomplete a string:
+```python
+from pyalex import Autocompletes
+
+Autocompletes()["stockholm"]
````
```suggestion
from pyalex import autocomplete

autocomplete("stockholm")
```

My suggestion would be to rewrite it into this as autocomplete is not a separate resource (although the OpenAlex API design might let users think this). 

With this change, we reserve the item retrieval for identifiers.

J535D165 commented on README.md:
````diff
@@ -307,6 +319,32 @@ for page in pager:
+```python
+from pyalex import Works
+
+Works().filter(cited_by_count=">1000", publication_year=2009).autocomplete("planetary boundaries")
````
Very nice!

J535D165 commented on pyalex/api.py:
```diff
@@ -169,8 +169,12 @@ def __init__(self, params=None):
 
-    def _full_collection_name(self):
-        return config.openalex_url + "/" + self.__class__.__name__.lower()
+    def _full_collection_name(self, autocomplete=False):
+        if autocomplete:
+            base_url = config.openalex_url + "/autocomplete/"
+            return base_url + self.__class__.__name__.lower()
+        else:
+            return config.openalex_url + "/" + self.__class__.__name__.lower()
```
I think we can simplify this and keep it unchanged.

romain894 replied:
We could replace `if autocomplete:` by `if 'q' in self.params.keys():`

J535D165 commented on pyalex/api.py:
```diff
@@ -197,7 +201,13 @@ def __getitem__(self, record_id):
-            return self._full_collection_name()
+            return self._full_collection_name(autocomplete=False)
+
+        if 'q' in self.params.keys():
+            # if q is in params, then it means that an autocomplete query was asked
+            autocomplete = True
+        else:
+            autocomplete = False
```
Same here

romain894 replied:
And delete this part

J535D165 commented on pyalex/api.py:
```diff
@@ -321,6 +331,11 @@ def select(self, s):
         self._add_params("select", s)
         return self
 
+    def autocomplete(self, s, **kwargs):
+        """ autocomplete the string s, for a specific type of entity """
+        self._add_params("q", s)
+        return self.get(**kwargs)
```
My suggestion would be to implement it like this:

```suggestion
    def autocomplete(self, q, return_meta=False):
        """Autocomplete query q for entity"""

        self._add_params("q", q)

        # manipulate URL for autocomplete
        url_split = urlsplit(self.url)
        new_url = urlunsplit(
            (
                url_split.scheme,
                url_split.netloc,
                f"/autocomplete{url_split.path}",
                url_split.query,
                url_split.fragment,
            )
        )

        return self._get_from_url(
            new_url, return_meta=return_meta, resource_class=Autocomplete
        )
```

Also, it's not the nicest option, but it keeps the autocomplete functionality separated from everything else. We don't need the changes above. As I don't think this was the most intuitive implementation OpenAlex chose, we might benefit from keeping it separate from the rest of the class. 

I changed `_get_from_url` to accept an extra argument.

J535D165 commented on pyalex/api.py:
```diff
@@ -401,6 +416,20 @@ class Funders(BaseOpenAlex):
+    pass
+
+
+class Autocompletes(BaseOpenAlex):
+    """ Class to autocomplete without being based on the type of entity """
+    resource_class = Autocomplete
+
+    def __getitem__(self, key):
+        return self._get_from_url(
+            config.openalex_url + "/autocomplete" + "?q=" + key, return_meta=False
+        )
```
Ideally, we can make this into something like this:

```python
def autocomplete(q, return_meta=False):
    BaseOpenAlex().autocomplete(q, return_meta=return_meta)
```

I expect this to require some changes to the base class as well.

J535D165 reviewed (changes requested):
Thanks a lot for the PR and the documentation! It is pleasure to have a feature-complete library soon :) 

I never took the effort to implement it, as, in my opinion, the design of the autocomplete API of OpenAlex is not very intuitive (autocomplete looks like a resource instead of a method). I'm happy with your abstraction autocomplete. This makes things more straightforward. I have a couple of remarks on your implementation. 

It might also be good to add a unit test. Something like this:

```python
def test_autocomplete():
    w = Works().autocomplete("open ale")

    assert not isinstance(w[0], Work)
```

romain894 commented on pyalex/api.py:
```diff
@@ -212,9 +222,10 @@ def url(self):
         if l_params:
-            return self._full_collection_name() + "?" + "&".join(l_params)
+            return self._full_collection_name(
+                autocomplete=autocomplete) + "?" + "&".join(l_params)
```
remove `autocomplete=autocomplete`

romain894 commented on pyalex/api.py:
```diff
@@ -212,9 +222,10 @@ def url(self):
+                autocomplete=autocomplete) + "?" + "&".join(l_params)
 
-        return self._full_collection_name()
+        return self._full_collection_name(autocomplete=autocomplete)
```
same here

romain894 commented:
The first reason why I implemented the autocomplete as the other entities was to benefit from the already existing functions of the library like paging. I just did some tests and realized it's not implemented on the OpenAlex API. The only parameters for autocomplete are `q, filter, mailto, search, hide_zero, author_hint`.

The other reason was to reuse as much code as possible but as you shown this may not be the only solution. I'm still not sure whether keeping my changes in `_full_collection_name` or your solution to rewriting `autocomplete` is the best.

From my perspective, managing the autocomplete in `_full_collection_name` and keeping the same structure as the other entities is easier and needs less code but I might have some bias as I'm the author of this code.

romain894 commented:
@J535D165 I implemented the changes and added 2 tests. The changes made passed the tests with python 3.10 on my computer (GitHub seems to have trouble running the tests, as the error 429 too many requests occur)

J535D165 commented:
Thanks @romain894. Planning on a new release soon!
